module/zfs/vfs_ratelimit.c

```diff
@@ -0,0 +1,664 @@
+		    slot->rts_timeslot == now) {
+			return (slot);
+		}
+	}
+	/* All the entries expired. */
+#ifndef NDEBUG
+	for (slot = list_head(&rl->rl_list); slot != NULL;
+	    slot = list_next(&rl->rl_list, slot)) {
+		ASSERT(slot->rts_timeslot < now);
+	}
+#endif
+
+	return (NULL);
+}
+
+/*
+ * Account for our request across all the types configured in this ratelimit
+ * structure.
+ * Return a timeslot we should wait for or now if we can execute the request
+ * without waiting (we are within limits).
+ */
+static uint64_t
+ratelimit_account(struct vfs_ratelimit *rl, int types, hrtime_t now,
+    const uint64_t *counts)
+{
+	uint64_t timeslot;
+	int type, typebit;
+
+	timeslot = 0;
+
+	mutex_enter(&rl->rl_lock);
+
+	for (type = ZFS_RATELIMIT_FIRST; type <= ZFS_RATELIMIT_LAST; type++) {
+		struct rtslot *slot;
+		uint64_t count, nexttimeslot;
+
+		typebit = (1 << type);
+
+		if ((types & typebit) == 0) {
+			/* Not interested in this type. */
+			continue;
+		}
+		if (rl->rl_limits[type] == 0) {
+			/* This type has no limit configured on this dataset. */
+			continue;
+		}
+		count = counts[type];
+		ASSERT(count > 0);
+
+		slot = ratelimit_find(rl, typebit, now);
+		if (slot == NULL) {
+			slot = list_remove_tail(&rl->rl_list);
+			ASSERT(slot->rts_timeslot < now);
+			slot->rts_types = typebit;
+			slot->rts_timeslot = now;
+			memset(slot->rts_counts, 0, sizeof (slot->rts_counts));
+			list_insert_head(&rl->rl_list, slot);
+		} else if (slot->rts_timeslot == now) {
+			/* The 'now' slot may not have our type yet. */
+			slot->rts_types |= typebit;
+		}
+		ASSERT((slot->rts_types & typebit) != 0);
+		nexttimeslot = slot->rts_timeslot + 1;
+
+		for (;;) {
+			if (slot->rts_counts[type] + count <=
+			    rl->rl_limits[type]) {
+				slot->rts_counts[type] += count;
+				break;
+			}
+
+			/*
+			 * This request is too big to fit into a single slot,
+			 * ie. a single request exceeds the limit or this and
+			 * the previous requests exceed the limit.
+			 */
+
+			/*
+			 * Fit as much as we can into the current slot.
+			 */
+			count -= rl->rl_limits[type] - slot->rts_counts[type];
+			slot->rts_counts[type] = rl->rl_limits[type];
+
+			/*
+			 * Take the next slot (if already exists isn't aware of
+			 * our type yet), take an expired slot from the tail of
+			 * the list or allocate a new slot.
+			 */
+			slot = list_prev(&rl->rl_list, slot);
+			if (slot != NULL) {
+				ASSERT((slot->rts_types & typebit) == 0);
+				ASSERT(slot->rts_timeslot == nexttimeslot);
+				ASSERT0(slot->rts_counts[type]);
+
+				slot->rts_types |= typebit;
+			} else {
+				slot = list_tail(&rl->rl_list);
+				if (slot->rts_timeslot < now) {
+					list_remove(&rl->rl_list, slot);
+				} else {
+					slot = kmem_alloc(sizeof (*slot),
+					    KM_SLEEP);
+				}
+				slot->rts_types = typebit;
+				slot->rts_timeslot = nexttimeslot;
+				memset(slot->rts_counts, 0,
+				    sizeof (slot->rts_counts));
+				list_insert_head(&rl->rl_list, slot);
```

Comment:
So if somebody ask for 16MB I/O with a bandwidth limit set to 16B/s you will allocate a million of slots here and they will stay with the dataset till eviction?  Why instead not just create 6 static "slots" (one per type) and just track there time till which specific type is busy and remaining count after that, if we limit time resolution, as it is done now?

Reply:
I've implemented something like that, no more slots and memory allocations.